src/communication/v2/server.hpp

```diff
@@ -0,0 +1,122 @@
+#include "communication/v2/listener.hpp"
+#include "communication/v2/pool.hpp"
+#include "utils/logging.hpp"
+#include "utils/message.hpp"
+#include "utils/thread.hpp"
+
+namespace memgraph::communication::v2 {
+
+using Socket = boost::asio::ip::tcp::socket;
+using ServerEndpoint = boost::asio::ip::tcp::endpoint;
+/**
+ * Communication server.
+ *
+ * Listens for incoming connections on the server port and assigns them to the
+ * connection listener. The listener and session are implemented using asio
+ * async model. Currently the implemented model is io_context per core model
+ * opposed to thread per core with global io_context. The only comparison
+ * found between those two is here (https://konradzemek.com/) and goes in favour of latter in the terms
+ * of scalability.
+ * All logic is contained within handlers that are being dispatched
+ * on a single strand per session. The only exception is write which is
+ * synchronous since the nature of the clients conenction is synchronous as
+ * well.
+ *
+ * Current Server architecture:
+ * incoming connection -> server -> listener -> session
+
+ *
+ * @tparam TSession the server can handle different Sessions, each session
+ *         represents a different protocol so the same network infrastructure
+ *         can be used for handling different protocols
+ * @tparam TSessionData the class with objects that will be forwarded to the
+ *         session
+ */
+template <typename TSession, typename TSessionData>
+class Server final {
+  using ServerHandler = Server<TSession, TSessionData>;
+
+ public:
+  /**
+   * Constructs and binds server to endpoint, operates on session data and
+   * invokes workers_count workers
+   */
+  Server(ServerEndpoint &endpoint, TSessionData *session_data, ServerContext *server_context,
+         const int inactivity_timeout_sec, const std::string_view service_name,
+         size_t workers_count = std::thread::hardware_concurrency())
+      : endpoint_{endpoint},
+        service_name_{service_name},
+        context_thread_pool_{workers_count},
+        listener_{Listener<TSession, TSessionData>::Create(context_thread_pool_, session_data, server_context,
+                                                           endpoint_, service_name_, inactivity_timeout_sec)} {}
+
+  ~Server() { MG_ASSERT(!IsRunning(), "Server wasn't shutdown properly"); }
+
+  Server(const Server &) = delete;
+  Server(Server &&) = delete;
+  Server &operator=(const Server &) = delete;
+  Server &operator=(Server &&) = delete;
+
+  const auto &Endpoint() const {
+    MG_ASSERT(IsRunning(), "You can't get the server endpoint when it's not running!");
+    return endpoint_;
+  }
+
+  bool Start() {
+    MG_ASSERT(!IsRunning(), "The server was already started!");
+    listener_->Start();
+
+    spdlog::info("{} server is fully armed and operational", service_name_);
+    spdlog::info("{} listening on {}", service_name_, endpoint_.address());
+    context_thread_pool_.Run();
+
+    return true;
```

Comment:
This `function` always returns `true` and never `false`. Consider making it `void` or handle the `false` case.